doc/cephfs/charmap.rst

```diff
@@ -0,0 +1,185 @@
+
+It can also be used to **remove** all settings and restore the default CephFS behavior
+for directory entry names (uninterpreted bytes without ``/`` that are NUL terminated).
+
+Note the following restriction on manipulating any of these extended attributes:
+
+* The directory must be empty.
+* The directory must not be part of a snapshot.
+
+New sub-directories created under a directory with a **charmap** configuration will
+inherit (copy) the parent's configuration.
+
+.. note:: You can remove a **charmap** on a sub-directory which inherited
+          the configuration so long as the preconditions apply: it is empty
+          and not part of an existing snapshot.
+
+
+Normalization
+-------------
+
+The ``ceph.dir.normalization`` configuration accepts the following normalization forms:
+
+* **nfd**: Form D (Canonical Decomposition)
+* **nfc**: Form C (Canonical Decomposition, followed by Canonical Composition)
+* **nfkd**: Form KD (Compatibility Decomposition)
+* **nfkc**: Form KC (Compatibility Decomposition, followed by Canonical Composition)
+
+The default normalization for a character mapping configuration is ``nfd``.
+
+.. note:: For more information about Unicode normalization forms, please see `Unicode normalization standard documents`_.
+
+Whenever a directory entry name is generated during path traversal or lookup,
+the client will apply the normalization to the name before submitting any
+operation to the MDS. On the MDS side, the directory entry names which
+are stored are only these normalized names.
+
+For example, to set the normalization on a directory:
+
+::
+
+    $ setfattr -n ceph.dir.normalization -v "" foo/
+    
+    $ getfattr -n ceph.dir.charmap foo/
+    # file: foo/
+    ceph.dir.charmap="{\"casesensitive\":true,\"normalization\":\"nfd\",\"encoding\":\"utf8\"}"
+    
+    $ getfattr -n ceph.dir.normalization foo/
+    # file: foo/
+    ceph.dir.normalization="nfd"
+
+.. note:: Setting the empty string will cause the MDS to pick the default normalization.
+
+All character mapping configurations must have a normalization enabled. Removing the normalization
+will cause the default to be restored:
+
+::
+
+    $ setfattr -n ceph.dir.normalization -v nfc foo/
+    $ getfattr -n ceph.dir.normalization foo/
+    # file: foo/
+    ceph.dir.normalization="nfc"
+    
+    $ setfattr -x ceph.dir.normalization foo/
+    $ getfattr -n ceph.dir.normalization foo/
+    # file: foo/
+    ceph.dir.normalization="nfd"
+
+To remove normlization on a directory, you must remove the ``ceph.dir.charmap``
+configuration.
+
+.. note:: The MDS maintains an ``alternate_name`` metadata (also used for
+          encryption) for directory entries which allows the client to persist the
+          original un-normalized name used by the application. The MDS does not
+          interpret this metadata in any way; it's only used by clients to reconstruct
+          the original name of the directory entry.
```

Comment:
suggest

>The MDS maintains ``alternate_name`` metadata (also used for encryption) for directory entries, which allows

`metadata` to me seems plural, which would conflict with `an`.  I think adding the above comma helps make more clear that the antecedent of`which` is `metadata`, not `entries`.

Also suggest : instead of ;

Reply:
I thought the singular is `datum`.

Reply:
https://www.merriam-webster.com/dictionary/data